File: python/src/linodemcp/cli/call.py
"""``linodemcp call <tool> [args]`` - run any registered tool from the shell.

The whole point is the one architectural rule: this never touches a tool
handler directly. It validates the tool name against the registry, builds an
arguments dict (from ``--json`` or ``--arg`` plus the safety flags), and feeds
it to ``Server.dispatch``. Every audit, profile-filter, dry-run, and two-stage
behavior an MCP request gets, a ``call`` invocation gets too, because it is the
same chokepoint.

Argument plumbing and result rendering live in ``_shared``; the runtime
(config load, server build, audit sink) lives in ``runtime``. This module is
the argparse surface and the exit-code policy.
"""

from __future__ import annotations

import argparse
from typing import TYPE_CHECKING, Any, TextIO

from linodemcp.cli._shared import (
    EXIT_SUCCESS,
    EXIT_TOOL_ERROR,
    EXIT_USAGE_ERROR,
    ArgError,
    build_arguments,
    is_error_result,
    render_output,
    result_text,
)
from linodemcp.cli.runtime import OneShotRuntime, open_runtime
from linodemcp.server import get_tool_registry

if TYPE_CHECKING:
    from pathlib import Path

    from mcp.types import Tool

_OUTPUT_CHOICES = ("json", "table")
_MODE_CHOICES = ("plan", "apply")


def _build_parser() -> argparse.ArgumentParser:
    """Construct the argparse parser for ``call``.

    ``--json`` and repeated ``--arg`` carry the tool arguments; they are
    mutually exclusive (enforced after parse so the error is a usage exit, not
    an argparse SystemExit). The safety flags mirror the MCP fields one to one.
    """
    parser = argparse.ArgumentParser(
        prog="linodemcp call",
        description="Run a registered tool non-interactively.",
        add_help=True,
    )
    parser.add_argument("tool", help="tool name (see `linodemcp tools`)")
    parser.add_argument(
        "--json",
        dest="json_args",
        metavar="OBJECT",
        help="arguments as a JSON object, e.g. '{\"instance_id\":123}'",
    )
    parser.add_argument(
        "--arg",
        dest="arg_pairs",
        action="append",
        default=[],
        metavar="KEY=VALUE",
        help="one argument, repeatable; typed by the tool schema",
    )
    parser.add_argument(
        "--output",
        choices=_OUTPUT_CHOICES,
        default="json",
        help="result rendering (default: json)",
    )
    _add_safety_flags(parser)
    return parser


def _add_safety_flags(parser: argparse.ArgumentParser) -> None:
    """Add the write-safety flags that fold into the arguments dict."""
    parser.add_argument("--dry-run", action="store_true", help="preview, no change")
    parser.add_argument("--confirm", action="store_true", help="confirm a mutation")
    parser.add_argument(
        "--mode",
        choices=_MODE_CHOICES,
        default=None,
        help="two-stage write mode",
    )
    parser.add_argument("--plan-id", default=None, help="plan id for mode=apply")
    parser.add_argument(
        "--confirmed-dry-run",
        action="store_true",
        help="assert a prior dry-run for a destroy",
    )
    parser.add_argument(
        "--yolo",
        action="store_true",
        help="bypass gating (only if the profile allows)",
    )
    parser.add_argument("--environment", default=None, help="named environment")


def _find_tool(name: str) -> Tool | None:
    """Look up a tool definition by name in the full registry.

    Uses the full registry (not the profile-filtered set) so an unknown name
    and a profile-filtered name get different, accurate messages: dispatch
    refuses a filtered tool, but the name itself is real.
    """
    for entry in get_tool_registry():
        if entry.name == name:
            return entry.tool
    return None


def callable_tool_names() -> frozenset[str]:
    """Return the set of tool names ``call`` can invoke.

    This is exactly the registry: ``call`` keeps no allow list of its own, so a
    name is callable iff it is registered. Public so the CLI-to-registry parity
    test can assert the surface matches without reaching into private lookups,
    mirroring the tool-manifest gate.
    """
    return frozenset(entry.name for entry in get_tool_registry())


def _known_tool_names() -> list[str]:
    """Sorted registry tool names, for the unknown-tool error hint."""
    return sorted(callable_tool_names())


def _parse_and_validate(
    argv: list[str],
    stderr: TextIO,
) -> tuple[argparse.Namespace, Tool] | None:
    """Parse argv and resolve the tool, or print a usage error and return None.

    Returns the parsed namespace plus the resolved ``Tool`` on success. Returns
    None for any pre-dispatch problem (argparse error, unknown tool, both arg
    sources given); the caller maps None to ``EXIT_USAGE_ERROR``.
    """
    parser = _build_parser()
    try:
        ns = parser.parse_args(argv)
    except SystemExit:
        # argparse already wrote its own message to stderr.
        return None

    if ns.json_args is not None and ns.arg_pairs:
        print("--json and --arg are mutually exclusive", file=stderr)
        return None

    tool = _find_tool(ns.tool)
    if tool is None:
        print(f"unknown tool: {ns.tool}", file=stderr)
        print(_unknown_tool_hint(ns.tool), file=stderr)
        return None

    return ns, tool


def _unknown_tool_hint(name: str) -> str:
    """Build a recovery hint for an unknown tool name.

    Suggests registry names that share the unknown name's prefix segment (the
    scope, e.g. ``linode_instance``) so a typo in the verb is easy to spot;
    falls back to the catalog pointer when nothing is close.
    """
    prefix = name.rsplit("_", 1)[0] if "_" in name else name
    near = [n for n in _known_tool_names() if n.startswith(prefix)]
    if near:
        listed = ", ".join(near[:10])
        return f"did you mean one of: {listed}"
    return "Run `linodemcp tools --all` to list every tool."


def _assemble_arguments(
    ns: argparse.Namespace,
    tool: Tool,
    stderr: TextIO,
) -> dict[str, Any] | None:
    """Build the dispatch arguments dict from the namespace, or None on error.

    Wraps ``build_arguments`` so an ``ArgError`` (bad json, bad key=value, a
    value that does not fit its schema type) becomes a stderr message and a
    None return the caller maps to ``EXIT_USAGE_ERROR``.
    """
    try:
        return build_arguments(
            tool,
            json_args=ns.json_args,
            arg_pairs=ns.arg_pairs,
            dry_run=ns.dry_run,
            confirm=ns.confirm,
            mode=ns.mode,
            plan_id=ns.plan_id,
            confirmed_dry_run=ns.confirmed_dry_run,
            yolo=ns.yolo,
            environment=ns.environment,
        )
    except ArgError as exc:
        print(str(exc), file=stderr)
        return None


async def run_call_command(
    argv: list[str],
    stdout: TextIO,
    stderr: TextIO,
    config_path: Path | None = None,
) -> int:
    """Entry point for ``linodemcp call``. Async: dispatch is async.

    Validates and assembles before building the runtime so a usage error never
    pays the server-construction cost. ``config_path`` overrides the standard
    config path (tests pass a temp file). Output streams are parameters so
    tests assert on captured text and the exit code.
    """
    parsed = _parse_and_validate(argv, stderr)
    if parsed is None:
        return EXIT_USAGE_ERROR
    ns, tool = parsed

    arguments = _assemble_arguments(ns, tool, stderr)
    if arguments is None:
        return EXIT_USAGE_ERROR

    return await _dispatch_call(
        ns.tool, arguments, ns.output, stdout, stderr, config_path
    )


async def _dispatch_call(
    name: str,
    arguments: dict[str, Any],
    output: str,
    stdout: TextIO,
    stderr: TextIO,
    config_path: Path | None,
) -> int:
    """Open the runtime, dispatch the call, render, and pick the exit code.

    A construction failure (bad config, unknown active profile) prints to
    stderr and exits 2. A tool-level error result prints its message to stderr
    and the raw payload to stdout, exiting 1. Success prints the rendered
    payload and exits 0.
    """
    try:
        async with open_runtime(config_path) as runtime:
            return await _run_and_render(
                runtime, name, arguments, output, stdout, stderr
            )
    except Exception as exc:
        print(f"failed to start runtime: {exc}", file=stderr)
        return EXIT_USAGE_ERROR


async def _run_and_render(
    runtime: OneShotRuntime,
    name: str,
    arguments: dict[str, Any],
    output: str,
    stdout: TextIO,
    stderr: TextIO,
) -> int:
    """Dispatch one call through the runtime's server and render the result.

    A ``ValueError`` from dispatch is the refused-tool path (profile filtered
    the tool out, or an unknown name slipped past validation): exit 2. Anything
    else propagating is an unexpected handler crash: surface it and exit 1.
    """
    try:
        result = await runtime.server.dispatch(name, arguments)
    except ValueError as exc:
        print(str(exc), file=stderr)
        return EXIT_USAGE_ERROR
    except Exception as exc:
        print(f"tool dispatch failed: {exc}", file=stderr)
        return EXIT_TOOL_ERROR

    text = result_text(result)
    if is_error_result(text):
        print(text, file=stderr)
        print(text, file=stdout)
        return EXIT_TOOL_ERROR

    print(render_output(text, output), file=stdout)
    return EXIT_SUCCESS

File: go/internal/cli/call.go
package cli

import (
	"context"
	"flag"
	"io"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/server"
)

// stringList is a flag.Value that accumulates repeated --arg occurrences
// into a slice, so `--arg a=1 --arg b=2` yields two entries.
type stringList []string

// String renders the accumulated values for flag's usage output.
func (*stringList) String() string {
	return ""
}

// Set appends one occurrence's value, called once per flag instance.
func (s *stringList) Set(value string) error {
	*s = append(*s, value)

	return nil
}

const callUsage = `Usage: linodemcp call <tool> [flags]

Run one tool through the same dispatch the MCP server uses: the call is
validated, audited, and subject to profile, dry-run, and two-stage rules
exactly as an MCP client's call would be.

Arguments (mutually exclusive):
  --arg key=value     Set one argument; repeatable. Values are typed from
                      the tool's schema (number, integer, boolean, string).
  --json '{...}'      Supply all arguments as one JSON object.

Safety flags (folded into the request like the MCP fields):
  --dry-run           Preview without executing (sets dry_run).
  --confirm           Confirm a write/destroy (sets confirm).
  --mode plan|apply   Two-stage destroy mode.
  --plan-id <id>      Plan id from a prior --mode plan, for --mode apply.
  --confirmed-dry-run Acknowledge a dry-run preview before a real run.
  --yolo              Bypass confirm where the profile allows it.
  --environment <name> Target a named environment.

Output:
  --output json       Print the tool's result payload (default).
  --output table      Render a list/object result as a text table.

Discover tools with "linodemcp tools" and "linodemcp tools show <tool>".`

// RunCallCommand runs `linodemcp call <tool> [flags]` and returns the
// process exit code: 0 on success, 1 when the tool returns an error
// result, 2 for any usage problem (no tool, unknown tool, bad flags,
// bad JSON, mutually-exclusive args). Output streams are parameters so
// tests capture stdout and stderr.
//
// The command builds a tools/call request and drives it through the
// server's HandleMessage. It reimplements no tool logic: validation,
// the confirm gate, dry-run, two-stage, audit, and profile filtering all
// happen inside the dispatch.
func RunCallCommand(args []string, stdout, stderr io.Writer) int {
	parsed, code, done := parseCallArgs(args, stderr)
	if done {
		return code
	}

	ctx := context.Background()

	runtime, err := newRuntime(ctx, stderr)
	if err != nil {
		writef(stderr, "%v\n", err)

		return 1
	}
	defer runtime.Close()

	if exit, handled := rejectUnknownTool(runtime.Server, parsed.tool, stderr); handled {
		return exit
	}

	schema := schemaForTool(runtime.Server, parsed.tool)

	arguments, err := BuildArguments(schema, parsed.jsonArg, parsed.kvArgs)
	if err != nil {
		writef(stderr, "%v\n", err)

		return ExitUsageError
	}

	ApplySafetyFlags(arguments, &parsed.safety)

	return executeCall(ctx, runtime.Server, &parsed, arguments, stdout, stderr)
}

// parsedCall holds the flag-parse output for a call invocation.
type parsedCall struct {
	tool    string
	jsonArg string
	kvArgs  []string
	output  string
	safety  SafetyFlags
}

// parseCallArgs parses the call flags. The third return reports whether
// the caller should stop (true) and use the returned exit code, which
// covers both the no-tool usage error and a -h/--help request.
func parseCallArgs(args []string, stderr io.Writer) (parsedCall, int, bool) {
	if len(args) == 0 {
		writeln(stderr, callUsage)

		return parsedCall{}, ExitUsageError, true
	}

	tool := args[0]

	flags := flag.NewFlagSet("call", flag.ContinueOnError)
	flags.SetOutput(stderr)
	flags.Usage = func() { writeln(stderr, callUsage) }

	var (
		kvArgs stringList
		parsed parsedCall
	)

	flags.StringVar(&parsed.jsonArg, "json", "", "all arguments as one JSON object")
	flags.Var(&kvArgs, "arg", "one argument as key=value (repeatable)")
	flags.StringVar(&parsed.output, "output", outputJSON, "output format: json or table")
	bindSafetyFlags(flags, &parsed.safety)

	if err := flags.Parse(args[1:]); err != nil {
		// flag already printed the error and usage; classify as usage.
		return parsedCall{}, ExitUsageError, true
	}

	parsed.tool = tool
	parsed.kvArgs = kvArgs

	if !validOutput(parsed.output) {
		writef(stderr, "unknown --output %q: want json or table\n", parsed.output)

		return parsedCall{}, ExitUsageError, true
	}

	return parsed, 0, false
}

// bindSafetyFlags registers the safety flags on a flag set, wiring the
// bool ones through tri-state pointers so an unset flag stays absent from
// the request. flag's BoolVar can't express "unset", so each bool is
// captured via a bool pointer that flag's Func sets only when present.
func bindSafetyFlags(flags *flag.FlagSet, safety *SafetyFlags) {
	bindOptionalBool(flags, "dry-run", &safety.DryRun)
	bindOptionalBool(flags, "confirm", &safety.Confirm)
	bindOptionalBool(flags, "confirmed-dry-run", &safety.ConfirmedDry)
	bindOptionalBool(flags, "yolo", &safety.Yolo)

	flags.StringVar(&safety.Mode, "mode", "", "two-stage mode: plan or apply")
	flags.StringVar(&safety.PlanID, "plan-id", "", "plan id for --mode apply")
	flags.StringVar(&safety.Environment, "environment", "", "target environment name")
}

// bindOptionalBool registers a boolean flag whose presence is tracked.
// When the flag appears the target pointer is set so ApplySafetyFlags
// writes the field; when it is absent the pointer stays nil and the field
// is left off the request, so a tool's own default applies.
//
// flag.BoolFunc passes "true" for the bare --flag form and the literal
// value for --flag=<value>, so parseBoolFlag handles both and rejects a
// non-boolean value as a usage error.
func bindOptionalBool(flags *flag.FlagSet, name string, target **bool) {
	flags.BoolFunc(name, "", func(raw string) error {
		value, err := parseBoolFlag(raw)
		if err != nil {
			return err
		}

		*target = &value

		return nil
	})
}

// executeCall dispatches the request, prints the payload, and maps the
// result to an exit code. A transport/RPC error is exit 1 with the error
// on stderr; a tool IsError result is exit 1 with the message on stderr
// and the payload still printed to stdout; success is exit 0.
func executeCall(
	ctx context.Context,
	srv *server.Server,
	parsed *parsedCall,
	arguments map[string]any,
	stdout, stderr io.Writer,
) int {
	result, err := dispatchCall(ctx, srv, parsed.tool, arguments)
	if err != nil {
		writef(stderr, "%v\n", err)

		return 1
	}

	renderOutput(stdout, result.Text, parsed.output)

	if result.IsError {
		writeln(stderr, "tool returned an error result")

		return 1
	}

	return 0
}

// rejectUnknownTool checks the requested tool against the full catalog
// and, when it is not a registered tool, prints a helpful message and
// reports handled=true with exit code 2. A tool that exists but is hidden
// by the active profile is reported distinctly so the user knows to
// switch profiles rather than think the tool doesn't exist.
func rejectUnknownTool(srv *server.Server, tool string, stderr io.Writer) (int, bool) {
	for _, info := range srv.AllToolInfos() {
		if info.Name == tool {
			return 0, false
		}
	}

	writef(stderr, "unknown tool %q.\n", tool)
	writeln(stderr, `Run "linodemcp tools --all" to list every tool.`)

	return ExitUsageError, true
}

// schemaForTool returns the input schema for tool from the active-profile
// view, falling back to the full catalog. The active view is preferred so
// coercion reflects what this profile exposes, but a tool present only in
// the full catalog (allowed by rejectUnknownTool's broader check) still
// gets its schema for argument typing.
func schemaForTool(srv *server.Server, tool string) mcp.ToolInputSchema {
	for _, info := range srv.ToolInfos() {
		if info.Name == tool {
			return info.InputSchema
		}
	}

	for _, info := range srv.AllToolInfos() {
		if info.Name == tool {
			return info.InputSchema
		}
	}

	return mcp.ToolInputSchema{}
}

// validOutput reports whether the --output value is one the renderer
// understands.
func validOutput(format string) bool {
	return format == outputJSON || format == outputTable
}

// parseBoolFlag parses an explicit boolean flag value, returning a usage
// error for anything that isn't a recognized bool literal.
func parseBoolFlag(raw string) (bool, error) {
	switch raw {
	case "true", "1", "t", "T", "TRUE", "True":
		return true, nil
	case "false", "0", "f", "F", "FALSE", "False":
		return false, nil
	default:
		return false, errInvalidBool
	}
}
